File: CodingChallengeSolutionsC++/2563. Count the Number of Fair Pairs/2563.CounttheNumberofFairPairs.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static System.Runtime.InteropServices.JavaScript.JSType;

namespace CodingChallengeSolutions.CounttheNumberofFairPairs
{
    public class CounttheNumberofFairPairs
    {
        public long CountFairPairs(int[] nums, int lower, int upper)
        {
            // Brute Force -> for each value compare to all other values of higher index (O(n^2))
            // int size = nums.Count();
            //long pairCounter = 0 
            //for(int x = 0; x <size; ++x)
            //{
            //     for(int y = x+1; y <size; ++y)
            //    {
            //        int sum = nums[x]+nums[y];
            //        if(sum >= lower && sum <= upper)
            //        {
            //            ++pairCounter;
            //        }
            //    }
            //}

            // Two Pointer -> Sort array. Move in the upper and lower bounds to find all possible pairs. If the lower number plus higher number
            //                is too high, then any subsequent number will also be too high. So we can keep the uppe pointer where it is 
            // Big O - O(nLogn) where n is the size of the array

            Array.Sort(nums);
            return CountPairs(nums, upper) - CountPairs(nums, lower - 1);
        }

        private long CountPairs(int[] nums, int bound)
        {
            long count = 0;
            int lower = 0;
            int upper = nums.Count() - 1;

            while (lower < upper)
            {
                if (nums[lower] + nums[upper] > bound)
                {
                    --upper;
                }
                else
                {
                    count += upper - lower;
                    lower++;
                }
            }
            return count;
        }
    }
}

File: CodingChallengeSolutionsC++/2563. Count the Number of Fair Pairs/2563.CounttheNumberofFairPairs.cpp
#include <vector>
#include <algorithm>

class CounttheNumberofFairPairs{
public:
    long long countFairPairs(std::vector<int>& nums, int lower, int upper) {
        int size = nums.size();
        long long pairCounter = 0;

        // Brute Force -> for each value compare to all other values of higher index (O(n^2))
        //for(int x = 0; x <size; ++x)
        //{
        //     for(int y = x+1; y <size; ++y)
        //    {
        //        int sum = nums[x]+nums[y];
        //        if(sum >= lower && sum <= upper)
        //        {
        //            ++pairCounter;
        //        }
        //    }
        //}

        // Binary Search -> Sort the array, find what the lowest and highest value can be by using lower and upper bounds minus value at index 
        //                  Find the index of the lower and upper bounds (closest value less than lower lowest possible value, and same with highest)
        //                  Subtract the difference, that will give the number of pairs that will be between the lower and upper bounds

        std::sort(nums.begin(), nums.end());
        for (int x = 0; x < size; ++x)
        {
            int min = lower - nums[x];
            int max = upper - nums[x];
            auto low = std::lower_bound(nums.begin() + x + 1, nums.end(), min);
            auto high = std::upper_bound(nums.begin() + x + 1, nums.end(), max);
            pairCounter += high - low;
        }


        return pairCounter;

    }
};